CQLCookbook.md

```diff
@@ -0,0 +1,349 @@
+```
+*Note:* Value set Object Identifier Codes (OIDs) can be found through the [Value Set Authority Center (VSAC)](https://vsac.nlm.nih.gov/authoring).
+
+### D) Codes
+Define codes from codesystems (defined above) that can be called upon later in the library.
+```sql
+code "Hematocrit [Volume Fraction] of Arterial blood": '32354-3' from "LOINC"
+```
+
+### E) Request Parameter
+Define a request parameter which can be referenced anywhere in the CQL Libray.
+```sql
+parameter device_request DeviceRequest
+```
+*Note:* DRLS typically uses DeviceRequest, ServiceRequest and MedicationRequest as common parameters.
+
+### F) Patient Context
+Specify that the statements below this should be interpreted with reference to a single patient.
+```sql
+context Patient
+```
+
+### G) Define Statement
+The basic unit of logic within a library, a define statement introduces a named expression that can be referenced within the library, or by other libraries.
+This can operate like a query or like a function:
+```sql
+define RelevantDiagnoses: 
+  CodesFromConditions(Confirmed(ActiveOrRecurring([Condition: "Home Oxygen Therapy Qualifying Conditions"])))
+  
+define function HighestObservation(ObsList List<Observation>):
+  Max(ObsList O return NullSafeToQuantity(cast O.value as Quantity))
+```
+
+Define statements in CQL can be used to query specific information. Since most statements in DRLS prepopulation files come under the `context Patient` declaration, these statements are typically used to pull specific information from a patient's electronic health record (EHR). Declaration statements come in two forms: queries and functions.
+
+#### 'Query' Statements
+Define statements in CQL begin with the header:
+```sql
+define myStatement:
+//  Describe logic here
+```
+The header is succeeded by a flow of logical arguments that describe which specific information to pull from the patient's EHR. These type of statements can be called upon by other statemetents in the library, in order to filter out more specific patient information.
+
+#### 'Function' Statements
+Sometimes the logic defined in a CQL statement can either:
+1) Be quite complex
+2) Need to be reused in multiple 'define' statements throughout the CQL library
+
+In this case, it is typically advantageous to define a CQL function statement. Formatted very similarly to normal 'Define' statements, function statements have the following header:
+```sql
+define function myFunction(parameter1 parameter1_type, parameter2 parameter2_type, ...)
+//  Describe logic here
+```
+*Note: Function headers do not necesarily need to contain parameters.*
+
+The header is again succeeded by a flow of logical arguments. These arguments can then be accessed by simply calling the function elsewhere in the CQL Library.
+
+
+*Note:* See [Part 2](#2-drls-specific-statements) below for a list CQL define statements that are commonly used within DRLS.
+
+***
+
+## 2) DRLS Statement Templates
```

Comment:
I guess my question is about the purpose of the template. Is it supposed to include all the typical retrieval statements DRLS currently use?

Reply:
Essentially, yes. The goal was to provide snippets of commonly used statements that people who develop future rulesets can just copy and paste into their prepopulation files; the overarching goal is to ease the learning curve of writing the prepopulation files, and to standardize the rulesets a bit by hopefully having any future CQL files use the exact same functions.

My next task will be to work on developing a shared library of common CQL functions, so hopefully in the future a lot of the CQL snippets in here can be trimmed down to only include the common retrieval statements instead of also having to include the sometimes repetitive and cumbersome helper functions.